### PR TITLE
Edge refactoring to DAL and minor PIP improvements

### PR DESCRIPTION
**Pull request by VakarisZ**
# What is this? 
Fixes part of #662 

Add any further explanations here. 

## Checklist
* [x] Have you successfully tested your changes locally?
* [x] Is the TravisCI build passing? 

## Changes
Refactored edges to use DAL
Minor PIP improvements

### CODE REVIEW

File: monkey/monkey_island/cc/services/edge/displayed_edge.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            os = edge.scans[-1]["data"]["os"]
+
+        displayed_edge = DisplayedEdgeService.edge_to_net_edge(edge)
+
+        displayed_edge["ip_address"] = edge['ip_address']
+        displayed_edge["services"] = services
+        displayed_edge["os"] = os
+        # we need to deepcopy all mutable edge properties, because weak-reference link is made otherwise,
+        # which is destroyed after method is exited and causes an error later.
+        displayed_edge["exploits"] = deepcopy(edge['exploits'])
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/edge/displayed_edge.py:41]: Edge is a weak-reference link to mongoengine document. This link is destroyed once we exit the method where `edge = Edge.objects.get()`. I'm not sure how we can make this less error prone, maybe refactor `exploits` into embedded document...

File: monkey/monkey_island/cc/services/edge/edge.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+from monkey_island.cc.models.edge import Edge
+
+RIGHT_ARROW = "\u2192"
+
+
+class EdgeService:
+
+    @staticmethod
+    def get_or_create_edge(src_node_id, dst_node_id, src_label, dst_label):
+        edge = False
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/edge/edge.py:16]: I think it'll be better to start with `None` and in the `finally` use `if edge is None`

File: monkey/monkey_island/cc/services/edge/edge.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        try:
+            edge = Edge.objects.get(src_node_id=src_node_id, dst_node_id=dst_node_id)
+        except DoesNotExist:
+            edge = Edge(src_node_id=src_node_id, dst_node_id=dst_node_id)
+        finally:
+            if edge:
+                edge.src_label = src_label
+                edge.dst_label = dst_label
+                edge.save()
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/edge/edge.py:25]: Shouldn't we reuse update_label here?

File: monkey/monkey_island/cc/services/edge/edge.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        return edge
+
+    @staticmethod
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/edge/edge.py:28]: Why is this a static method and not a regular method of Edge?

File: monkey/monkey_island/cc/services/edge/edge.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        if edge.src_node_id == node_id:
+            edge.src_label = label
+        elif edge.dst_node_id == node_id:
+            edge.dst_label = label
+        else:
+            raise DoesNotExist("Node id provided does not match with any endpoint of an edge provided.")
+        edge.save()
+        pass
+
+    @staticmethod
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/edge/edge.py:39]: Why is this a static method and not a regular method of Edge?

File: monkey/monkey_island/cc/services/edge/edge.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                "data": machine_info
+            }
+        ip_address = machine_info.pop("ip_addr")
+        domain_name = machine_info.pop("domain_name")
+        edge.scans.append(new_scan)
+        edge.ip_address = ip_address
+        edge.domain_name = domain_name
+        edge.save()
+
+    @staticmethod
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/edge/edge.py:54]: Why is this a static method and not a regular method of Edge?

File: monkey/monkey_island/cc/services/edge/edge.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    def update_based_on_exploit(edge: Edge, exploit: Dict):
+        edge.exploits.append(exploit)
+        edge.save()
+        if exploit['result']:
+            EdgeService.set_edge_exploited(edge)
+
+    @staticmethod
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/edge/edge.py:61]: Why is this a static method and not a regular method of Edge?

File: monkey/monkey_island/cc/services/edge/edge.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    def set_edge_exploited(edge: Edge):
+        edge.exploited = True
+        edge.save()
+
+    @staticmethod
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/edge/edge.py:66]: Why is this a static method and not a regular method of Edge?

File: monkey/monkey_island/cc/services/edge/edge.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    def get_edge_group(edge: Edge):
+        if edge.exploited:
+            return "exploited"
+        if edge.tunnel:
+            return "tunnel"
+        if edge.scans or edge.exploits:
+            return "scan"
+        return "empty"
+
+    @staticmethod
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/edge/edge.py:76]: Why is this a static method and not a regular method of Edge?

File: monkey/monkey_island/cc/services/node.py [hunk 10]
```diff
@@ -186,23 +189,30 @@ def unset_all_monkey_tunnels(monkey_id):
...
-        mongo.db.edge.update(
-            {"from": monkey_id, 'tunnel': True},
-            {'$set': {'tunnel': False}},
-            upsert=False)
+        try:
+            edge = Edge.objects.get(src_node_id=monkey_id, tunnel=True)
+            edge.tunnel = False
+            edge.save()
+        except DoesNotExist:
+            pass
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/node.py:197]: Log a warning?

File: monkey/monkey_island/cc/services/edge/displayed_edge.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        edges = Edge.objects(dst_node_id=ObjectId(dst_id))
+        return [DisplayedEdgeService.edge_to_displayed_edge(edge, for_report) for edge in edges]
+
+    @staticmethod
+    def get_displayed_edge_by_id(edge_id, for_report=False):
+        edge = Edge.objects.get(id=edge_id)
+        displayed_edge = DisplayedEdgeService.edge_to_displayed_edge(edge, for_report)
+        return displayed_edge
+
+    @staticmethod
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/edge/displayed_edge.py:24]: Why is this a static method and not a regular method of Edge?

File: monkey/monkey_island/cc/services/edge/displayed_edge.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        return edge
+
+    @staticmethod
+    def services_to_displayed_services(services, for_report=False):
+        if for_report:
+            return [x for x in services]
+        else:
+            return [x + ": " + (services[x]['name'] if 'name' in services[x] else 'unknown') for x in services]
+
+    @staticmethod
```
Comment (ShayNehmad) [inline monkey/monkey_island/cc/services/edge/displayed_edge.py:66]: Why is this a static method and not a regular method of Edge?

File: monkey/monkey_island/cc/services/edge/displayed_edge.py (hunk 11 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/services/edge/displayed_edge.py:24]: I wanted to separate the data structure/schema of Edge document from methods, in order not to have a single huge file.  But you're right - we can do it without static method :) And I should make it into a boundary.

File: monkey/monkey_island/cc/services/node.py (hunk 10 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/services/node.py:197]: Not exactly, but solved

**Pull request merged**